native/dlahqr.go

```diff
@@ -0,0 +1,421 @@
+		}
+	}
+
+	// Quick return if possible.
+	if n == 0 {
+		return info
+	}
+	if ilo == ihi {
+		wr[ilo] = h[ilo*ldh+ilo]
+		wi[ilo] = 0
+		return info
+	}
+
+	// Clear out the trash.
+	for j := ilo; j < ihi-2; j++ {
+		h[(j+2)*ldh+j] = 0
+		h[(j+3)*ldh+j] = 0
+	}
+	if ilo <= ihi-2 {
+		h[ihi*ldh+ihi-2] = 0
+	}
+
+	nh := ihi - ilo + 1
+	nz := ihiz - iloz + 1
+
+	// Set machine-dependent constants for the stopping criterion.
+	ulp := dlamchP
+	smlnum := float64(nh) / ulp * dlamchS
+
+	// i1 and i2 are the indices of the first row and last column of H to
+	// which transformations must be applied. If eigenvalues only are being
+	// computed, i1 and i2 are set inside the main loop.
+	var i1, i2 int
+	if wantt {
+		i1 = 0
+		i2 = n - 1
+	}
+
+	itmax := 30 * max(10, nh) // Total number of QR iterations allowed.
+
+	// The main loop begins here. i is the loop index and decreases from ihi
+	// to ilo in steps of 1 or 2. Each iteration of the loop works with the
+	// active submatrix in rows and columns l to i. Eigenvalues i+1 to ihi
+	// have already converged. Either l = ilo or H[l,l-1] is negligible so
+	// that the matrix splits.
+	bi := blas64.Implementation()
+	i := ihi
+	for i >= ilo {
+		l := ilo
+
+		// Perform QR iterations on rows and columns ilo to i until a
+		// submatrix of order 1 or 2 splits off at the bottom because a
+		// subdiagonal element has become negligible.
+		converged := false
+		for its := 0; its < itmax; its++ {
+			// Look for a single small subdiagonal element.
+			var k int
+			for k = i; k >= l+1; k-- {
+				if math.Abs(h[k*ldh+k-1]) <= smlnum {
+					break
+				}
+				tst := math.Abs(h[(k-1)*ldh+k-1]) + math.Abs(h[k*ldh+k])
+				if tst == 0 {
+					if k-2 >= ilo {
+						tst += math.Abs(h[(k-1)*ldh+k-2])
+					}
+					if k+1 <= ihi {
+						tst += math.Abs(h[(k+1)*ldh+k])
+					}
+				}
+				// The following is a conservative small
+				// subdiagonal deflation criterion due to Ahues
+				// & Tisseur (LAWN 122, 1997). It has better
+				// mathematical foundation and improves accuracy
+				// in some cases.
+				if math.Abs(h[k*ldh+k-1]) <= ulp*tst {
+					ab := math.Max(math.Abs(h[k*ldh+k-1]), math.Abs(h[(k-1)*ldh+k]))
+					ba := math.Min(math.Abs(h[k*ldh+k-1]), math.Abs(h[(k-1)*ldh+k]))
+					aa := math.Max(math.Abs(h[k*ldh+k-0]), math.Abs(h[(k-1)*ldh+k-1]-h[k*ldh+k]))
+					bb := math.Min(math.Abs(h[k*ldh+k-0]), math.Abs(h[(k-1)*ldh+k-1]-h[k*ldh+k]))
+					s := aa + ab
+					if ab/s*ba <= math.Max(smlnum, aa/s*bb*ulp) {
+						break
+					}
+				}
+			}
+			l = k
+			if l > ilo {
+				// H[l,l-1] is negligible.
+				h[l*ldh+l-1] = 0
+			}
+			if l >= i-1 {
+				// Break the loop because a submatrix of order 1
+				// or 2 has split off.
+				converged = true
+				break
+			}
+
+			// Now the active submatrix is in rows and columns l to
+			// i. If eigenvalues only are being computed, only the
+			// active submatrix need be transformed.
+			if !wantt {
+				i1 = l
+				i2 = i
+			}
+
+			const (
```

Comment:
Ok, by going backwards and fitting the constants to the code and to the above references, I can get to something where these numbers come up, at least for the case of eispack and old lapack. 

A rough outline follows.

A shift polynomial is the characteristic polynomial of the 2x2 submatrix at the bottom right corner of the active block (the default, Francis' double shift, case here). If a 1x1 or 2x2 submatrix does not split off in 10 (and then in another 10) iterations, they use an exceptional shift polynomial `p_e` instead, for example,
`p_e(x) := x^2 - 3/2 s x + s^2`
where
`s := |h_{i,i-1}| + |h_{i-1,i-2}|.`
So they want to form a real 2x2 matrix `[h11 h12; h21 h22]` whose characteristic polynomial is `p_e`. They take the matrix in the form
`[ a b ]`
`[ b a ]`
Then its characteristic polynomial is
`det([ a-x b; b a-x]) = x^2 - 2ax + a^2-b^2`
Comparing coefficients in front of the same powers of `x` in `p_e`, we get
`3/2 s = 2 a`
`s^2 = a^2-b^2`
From the first equation we have that `a = 3/4 s`, and that's the 0.75 constant in, for example,
https://github.com/johannesgerer/jburkardt-f77/blob/master/eispack/eispack.f#L4389
Substitute `a` into the second equation and we get that `b^2 = -7/16 s^2`. And `-7/16 == -0.4375` in https://github.com/johannesgerer/jburkardt-f77/blob/master/eispack/eispack.f#L4391
But this means that b must be imaginary while they want a real matrix, so instead they form `[a b^2; b^2 a]`. I don't see an immediate justification as to why, maybe it will come later. Or maybe I'm missing something.

That's for the old lapack, the 10 and 20 cases in the switch here form the matrix in a different way. One possibility is that they use a different shift polynomial but I'm not sure.